libc/string_cpy.c

```diff
@@ -0,0 +1,1382 @@
+
+	char buff[MAX_ASCII_SET] = { 0 },
+		 ascii[MAX_ASCII_SET] = { 0 };
+	int i;
+
+	for (i = 1; i < MAX_ASCII_SET; i++) {
+		ascii[i - 1] = i;
+
+		/* This case test not only ascii set copied but different sizes of copy */
+		TEST_ASSERT_EQUAL_CHAR_ARRAY(ascii, strncpy(buff, ascii, i), i);
+	}
+
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(buff, ascii, sizeof(buff));
+
+#pragma GCC diagnostic pop
+}
+
+
+TEST(string_strncpy, append_null)
+{
+	char buff[MAX_STR_LEN] = { 0 };
+	char testStr[] = "Lorem \0 ipsum";
+
+	/* Checking if strncpy correctly read the string and if it is sensitivity on Null terminating zero */
+	TEST_ASSERT_EQUAL_STRING(testStr, strncpy(buff, testStr, sizeof(buff)));
+	TEST_ASSERT_NOT_EQUAL_CHAR(testStr[sizeof(testStr) / 2], buff[sizeof(testStr) / 2]);
+
+	memset(buff, 0, sizeof(buff));
+	memset(testStr, 0, sizeof(testStr));
+
+	/* Checking if we can copy place where NULL element is present */
+	TEST_ASSERT_EQUAL_PTR(buff, strncpy(buff, "", sizeof(buff)));
+	TEST_ASSERT_EQUAL_CHAR_ARRAY(buff, testStr, sizeof(testStr));
+}
+
+
+TEST(string_strncpy, adjacent)
+{
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+
+	char testStr[] = "TEST",
+		 memStr[MAX_STR_LEN] = "\0\0\0\0\0\0\0\0\0\0TEST",
+		 expVal[MAX_STR_LEN] = "\0\0\0\0\0\0TESTTEST",
+		 testStrLen = sizeof(testStr) - 1;
+
+	/* Copy data in same space but with offset to left */
+	TEST_ASSERT_EQUAL_STRING(&expVal[10 - testStrLen], strncpy(&memStr[10 - testStrLen], &memStr[10], testStrLen));
+
+	strcat(&expVal[10], testStr);
+
+	/* Copy data in same space but with offset to right */
+	TEST_ASSERT_EQUAL_STRING(&expVal[10 + testStrLen], strncpy(&memStr[10 + testStrLen], &memStr[10], testStrLen));
+
+#pragma GCC diagnostic pop
+}
+
+
+TEST(string_strncpy, one_byte)
+{
+	char buff[MAX_ASCII_SET] = { 0 },
+		 ascii[MAX_ASCII_SET] = { 0 };
+	int i;
+
+	/* Copy one by one from created ascii table */
+	for (i = 1; i < MAX_ASCII_SET; i++) {
+		ascii[i - 1] = i;
+
+		TEST_ASSERT_EQUAL_CHAR_ARRAY(buff, strncpy(buff, &ascii[i], 1), sizeof(buff));
+	}
+}
+
+
+TEST(string_strncpy, big)
+{
+	int num = 12345678;
+	char payload[BIG_NUMB] = { 0 },
+		 bigBuff[BIG_NUMB] = { 0 };
+
+	sprintf(payload, "%d", num);
+
+/* Pragma truncation disable for the length of string tests */
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wstringop-truncation"
+
+	/* Ability to copy long string */
+	TEST_ASSERT_EQUAL_STRING(bigBuff, strncpy(bigBuff, long_str, BIG_NUMB));
+	TEST_ASSERT_EQUAL_STRING(long_str, bigBuff);
+
+	/* Checking if a string doesn't contain any of the data after the size of input */
+	TEST_ASSERT_EQUAL_STRING(bigBuff, strncpy(bigBuff, payload, BIG_NUMB));
+	TEST_ASSERT_EQUAL_STRING(payload, bigBuff);
+	TEST_ASSERT_EQUAL_STRING(&bigBuff[BIG_NUMB / 2], "");
+	TEST_ASSERT_EQUAL_STRING(&bigBuff[BIG_NUMB / 3], "");
+	TEST_ASSERT_EQUAL_STRING(&bigBuff[BIG_NUMB / 4], "");
+	TEST_ASSERT_EQUAL_STRING(&bigBuff[BIG_NUMB / 5], "");
```

Comment:
That's checking the ability of appending NUL characters, but I would rather check all of the appended characters (and do it in the separate test case as mentioned somewhere above)